### PR TITLE
KIP-297: MINOR: move FileConfigProvider to provider subpackage

### PR DESCRIPTION
**Pull request by rayokota**
This moves FileConfigProvider to the org.apache.common.config.provider package to more easily isolate provider implementations going forward.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/common/config/FileConfigProviderTest.java [hunk 1]
```diff
@@ -16,6 +16,7 @@
  */
 package org.apache.kafka.common.config;
 
+import org.apache.kafka.common.config.provider.FileConfigProvider;
```
Comment (rhauch) [inline clients/src/test/java/org/apache/kafka/common/config/FileConfigProviderTest.java]: This `FileConfigProviderTest` should be moved into the `config.provider` package.

**Review by rhauch — APPROVED** [review]
LGTM, although the `FileConfigProviderTest` needs to be moved.

**Review by kkonstantine — CHANGES_REQUESTED** [review]
The interface belongs in the package of the providers. 
Should be `org.apache.kafka.common.config.provider.ConfigProvider`

See `Transformation` for a similar example.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigTransformer.java [hunk 2]
```diff
@@ -16,6 +16,8 @@
  */
 package org.apache.kafka.common.config;
```
Comment (kkonstantine) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigTransformer.java:17]: Should be `org.apache.kafka.common.config.provider` in a similar way we define interfaces in Connect. E.g. `package org.apache.kafka.connect.transforms.Transformation`.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java [hunk 3]
```diff
@@ -122,14 +122,16 @@
...
+    private static final String WHITELIST = "^org\\.apache\\.kafka\\.(?:connect\\.(?:"
             + "transforms\\.(?!Transformation$).*"
             + "|json\\..*"
             + "|file\\..*"
             + "|converters\\..*"
             + "|storage\\.StringConverter"
             + "|rest\\.basic\\.auth\\.extension\\.BasicAuthSecurityRestExtension"
-            + "))$";
+            + ")"
+            + "|common\\.config\\.provider\\..*"
```
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java]: Based on the above
`common\\.config\\.provider\\.(?!ConfigProvider$).*`

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java [hunk 4]
```diff
@@ -153,11 +156,14 @@ public void testAllowedConnectFrameworkClasses() throws Exception {
 
     @Test
     public void testClientConfigProvider() throws Exception {
+        assertFalse(PluginUtils.shouldLoadInIsolation(
+                "org.apache.kafka.common.config.ConfigProvider")
```
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java]: `org.apache.kafka.common.config.provider.ConfigProvider`

**Review by kkonstantine — COMMENTED** [review]
Pointed out the changes but we should change the description in the KIP as well here: https://cwiki.apache.org/confluence/display/KAFKA/KIP-297%3A+Externalizing+Secrets+for+Connect+Configurations

Additionally, changing package to `FileConfigProviderTest` should also be straightforward. Same for some tests that define a `TestConfigProvider`.

**Comment by rayokota** [comment]
@kkonstantine  has requested more changes, so https://github.com/apache/kafka/pull/5177 can be used to unblock the build for now

**Comment by rayokota** [comment]
retest this please

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java [hunk 5]
```diff
@@ -153,11 +156,14 @@ public void testAllowedConnectFrameworkClasses() throws Exception {
...
+        assertFalse(PluginUtils.shouldLoadInIsolation(
+                "org.apache.kafka.common.config.provider.ConfigProvider")
+        );
         assertTrue(PluginUtils.shouldLoadInIsolation(
-                "org.apache.kafka.common.config.FileConfigProvider")
+                "org.apache.kafka.common.config.provider.FileConfigProvider")
         );
         assertTrue(PluginUtils.shouldLoadInIsolation(
-                "org.apache.kafka.common.config.FutureConfigProvider")
+                "org.apache.kafka.common.config.provider.FutureConfigProvider")
```
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java:166]: nit: add also another case for something not ending in `ConfigProvider` ?

**Review by kkonstantine — APPROVED** [review]
Added an idea for another test case but, regardless, LGTM! Thanks @rayokota

**Comment by rayokota** [comment]
retest this please

**Comment by rayokota** [comment]
retest this please

**Comment by rayokota** [comment]
retest this please

**Review by ijuma — COMMENTED** [review]
Are the classes in the provider subpackage meant to be public API? If so, we need to add a javadoc include in the build file for the package.

**Comment by rayokota** [comment]
Thanks @ijuma , added the subpackage to the public JavaDoc

**Comment by rayokota** [comment]
retest this please

**Comment by rayokota** [comment]
retest this please

**Review by hachikuji — APPROVED** [review]
LGTM

**Pull request merged**